### ISSUE TITLE
Emit Outputing mapping is incorrect for AsofJoin 

### ISSUE THREAD

**Issue by icexelloss** [ibis-project/ibis-substrait#423]
Hi - I have found something wrong with Emit output mapping if I use asof join in my plan. The ibis-substrait what I used is in this branch: https://github.com/wjones127/ibis-substrait/tree/protobuf-3-19
Which includes a AsofJoin extension message:
https://github.com/wjones127/ibis-substrait/blob/protobuf-3-19/ibis_substrait/proto/extension_rels_pb2.py#L12

The code I have is internal, but it looks sth like this
```
t1 = ... #time, id, a
t2 = ... #time, id, b
joined = t1.asof_join(t2, on='time', by='id')
joined = joined[['time', 'id', 'a', 'b']]
```
When I execute the code with Acero, it crashes with a segment fault on this line
```
Thread 1 "/opt/ts/fuse/ar" received signal SIGSEGV, Segmentation fault.

arrow::engine::ProcessEmit<substrait::ProjectRel> (rel=..., no_emit_declr=..., schema=...) at /build/build/ext/public/apache/arrow/master/apache-arrow/cpp/src/arrow/engine/substrait/relation_internal.cc:78
```

After some debugging, it seems this is caused by the emit output mapping from the substrait plan, the emit mapping in my plan looks like this:

```
          emit {
            output_mapping: 6
            output_mapping: 7
            output_mapping: 8
            output_mapping: 9
          }
        }
```

After I hacked the ibis-substrait and forces it to produce the following emit mapping (based on information that @wjones127 told me), the query seems to work fine
```
          emit {
            output_mapping: 4
            output_mapping: 5
            output_mapping: 6
            output_mapping: 7
          }
        }
```

I am not sure how emit output mapping is supposed to work and where the problem is, please advice.

**Comment by gforsyth** [comment]
Can you post the full substrait plan?

**Comment by icexelloss** [comment]
```
 extension_uris {

  extension_uri_anchor: 1

}

relations {

  root {

    input {

      project {

        common {

          emit {

            output_mapping: 6

            output_mapping: 7

            output_mapping: 8

            output_mapping: 9

          }

        }

        input {

          extension_multi {

            inputs {

              read {

                base_schema {

                  names: "time"

                  names: "tid"

                  names: "return_raw_prev_1d_lag"

                  struct {

                    types {

                      timestamp_tz {

                        nullability: NULLABILITY_REQUIRED

                      }

                    }

                    types {

                      i32 {

                        nullability: NULLABILITY_NULLABLE

                      }

                    }

                    types {

                      fp64 {

                        nullability: NULLABILITY_NULLABLE

                      }

                    }

                    nullability: NULLABILITY_REQUIRED

                  }

                }

                named_table {

                  names: "smooth:/research/user/timothyd/bamboo/acero-model-state-wide/ms-19980101-20210301?begin=20200101&end=20200201&columns=tid,return_raw_prev_1d_lag"

                }

              }

            }

            inputs {

              read {

                base_schema {

                  names: "time"

                  names: "tid"

                  names: "CONDITIONER"

                  struct {

                    types {

                      timestamp_tz {

                        nullability: NULLABILITY_REQUIRED

                      }

                    }

                    types {

                      i32 {

                        nullability: NULLABILITY_NULLABLE

                      }

                    }

                    types {

                      fp64 {

                        nullability: NULLABILITY_NULLABLE

                      }

                    }

                    nullability: NULLABILITY_REQUIRED

                  }

                }

                named_table {

                  names: "smooth:/research/user/timothyd/bamboo/acero-model-state-wide/conditioner-19980101-20210301?begin=20200101&end=20200201&columns=tid,CONDITIONER"

                }

              }

            }

            detail {

              type_url: "type.googleapis.com/arrow.substrait_ext.AsOfJoinRel"

              value: "\n\022\n\006\022\004\n\002\022\000\022\010\022\006\n\004\022\002\010\001\n\022\n\006\022\004\n\002\022\000\022\010\022\006\n\004\022\002\010\001\020\200\200\244\311\377\301\211\001"

            }

          }

        }

        expressions {

          selection {

            direct_reference {

              struct_field {

              }

            }

            root_reference {

            }

          }

        }

        expressions {

          selection {

            direct_reference {

              struct_field {

                field: 1

              }

            }

            root_reference {

            }

          }

        }

        expressions {

          selection {

            direct_reference {

              struct_field {

                field: 2

              }

            }

            root_reference {

            }

          }

        }

        expressions {

          selection {

            direct_reference {

              struct_field {

                field: 3

              }

            }

            root_reference {

            }

          }

        }

      }

    }

    names: "time"

    names: "tid"

    names: "return_raw_prev_1d_lag"

    names: "CONDITIONER"
}
```

**Comment by icexelloss** [comment]
> so I would expect the 6 pre-existing columns to occupy indices [0, 5] and the four output columns to occupy [6,9]
The output schema of the join table is just "time, tid, return_raw_prev_1d_lag, CONDITIONER" and @wjones127 appears to think that something is wrong with the emit outputting. I don't understand this enough to say if this issue is on ibis-substrait or Acero. How exactly does "emit output mapping" works?

**Comment by wjones127** [comment]
There is a unit test within Arrow that contains the expected semantics for columns indices:

https://github.com/westonpace/arrow/blob/e8b99218ce09c23246a8320a1bd45790bcbb6df1/cpp/src/arrow/engine/substrait/serde_test.cc#L3665-L3815

In the plan it joins two tables:

```
left
----
0: time: i32
1: key: i32
2: value1: fp64

right
-----
0: time: i32
1: key: i32
2: value2: fp64
```

They are joined with `time` as `on` key and `key` as `by` key.

It them uses `"outputMapping": [0, 1, 2, 3]`, which outputs the columns `["time", "key", "value1", "value2"]`. This suggests that the columns used in the `on` and `by` key are "deduplicated".

I suggest you write a unit test that verifies your as-of join implementation can produce this  plan (or an equivalent one).

**Comment by icexelloss** [comment]
> I suggest you write a unit test that verifies your as-of join implementation can produce this plan (or an equivalent one).
I don't know how to do that, the emit output mapping is computed out side of my code (it is in translate.py) so I am not sure I can fix my code (if there is a way let me know)

**Comment by wjones127** [comment]
@icexelloss In the long-term, are you planning on contributing your as-of join implementation into this repo? If so, have you considered adding your as-of join implementation in a fork for now? That might make the output mapping logic easier to modify.

**Comment by icexelloss** [comment]
> In the long-term, are you planning on contributing your as-of join implementation into this repo? If so, have you considered adding your as-of join implementation in a fork for now? That might make the output mapping logic easier to modify.

I am not sure. Although I am not sure how that would help because the emit logic is computed here:
https://github.com/ibis-project/ibis-substrait/blob/main/ibis_substrait/compiler/translate.py#L744

One thing we can try to do is to make the ibis asof join node and I can just mimic that the ibis asof join node internally. (they are very similar)

**Comment by icexelloss** [comment]
But for the problem at hand - I don't think it is an issue with the translation rule for "asof join", if it is, let me know how I can fix it. Our implementation is already shared here:
https://github.com/ibis-project/ibis-substrait/issues/416#issue-1435219918

**Comment by wjones127** [comment]
So my understanding right now is that in order for you to be able to implement as-of join, we need to provide better utilities to handle the column indices for novel `translate()` implementations, such as this asof join case. @gforsyth does that make sense? I'd be happy to pair with you or someone else to get this working.

**Comment by gforsyth** [comment]
There are means of defining custom translation rules, I think the issue here is that the asof-join unpacking logic in Arrow seems to be in conflict with the existing rule for handling `ops.Selection`.
We can special case this behavior, but I would really rather not do that because it appears to me that the selection logic specifically around this node is different from other substrait selection logic.

**Comment by icexelloss** [comment]
> because it appears to me that the selection logic specifically around this node is different from other substrait selection logic.
Can you please elaborate?

**Comment by gforsyth** [comment]
For any other projection in substrait, so far as I am aware, something in this pattern:
```
t1 = ... #time, id, a
t2 = ... #time, id, b
joined = t1.join(t2, on...)
joined = joined[['time', 'id', 'a', 'b']]
```

will have a flattened list of columns consisting of
All columns in t1
All columns in t2
All columns in final projection

so

```
time
id
a
time
id
b
time
id
a
b
```

And to refer to those columns, you use the index of the flattened list to pull out the columns you want.  This is what `output_mapping` does.

From what I can see in the Arrow test and what @wjones127 said, the `asof_join` behavior is instead, given:

```
t1 = ... #time, id, a
t2 = ... #time, id, b
joined = t1.asof_join(t2, on=..., by=...)
joined = joined[['time', 'id', 'a', 'b']]
```

we get a flattened list of columns consisting of
All columns in t1 except if they've been used in an asof_join
All columns in t2 except if they've been used in an asof_join
All columns in final projection


So I could be missing something, but that just seems incorrect -- to generate a plan that's compliant with that we would walk up the expression tree, find all of the root tables, but then also check to see if any asof joins exist, then prune out any columns used in _certain fields_ of the asof join because of special casing, _then_ generate an output mapping.

**Comment by wjones127** [comment]
> So I could be missing something, but that just seems incorrect

Well it doesn't conform to the current expectations of ibis-substrait, but it's not clear to me the current setup is generally correct. Is there a place in the substrait spec we can point to that describes the correct behavior?

cc @westonpace

**Comment by westonpace** [comment]
> Is there a place in the substrait spec we can point to that describes the correct behavior?

No, asof join is an extension and not formally part of the spec.

I think we just need to agree on something.  @gforsyth is correct in that every other relation in Substrait (I think) includes all inputs in its output.  As an example, the join relation includes the key.

One could even argue that, in the asof join case, repeating the time column could be useful in some corner case scenarios (e.g. if there is tolerance then the keys don't exactly match and maybe that is useful in some situation).

So if I had to pick, off the cuff, I'd be in favor of aligning the C++ code to @gforsyth 's expectation.  That being said, I think changing the C++ code will potentially be more work.  The asof join isn't dropping those columns at the end.  It is dropping them somewhere in the middle (more specifically, when it materializes rows) and so we'd need to figure that out.

**Comment by westonpace** [comment]
I spoke with @gforsyth and I think we want to do this the right way.  So, if we agree it is in C++ I can probably get to it in the next week or two (I am a bit heads down at the moment trying to wrap up scanner tasks to unblock ARROW-15732).  If Li or Yaron have time to do a PR then I can review it as well.

**Comment by icexelloss** [comment]
```
every other relation in Substrait (I think) includes all inputs in its output. As an example, the join relation includes the key.
```
This is a bit surprising to me as SQL join operation (for example, in Spark) can de duplicate join keys as well. How does substrait handles that? Also, how about aggregation operation (aggregation doesn't include inputs in output as well)?

**Comment by icexelloss** [comment]
> One possible solution to this would be to leave the as-of node alone and only alter the Substrait->Acero mapping. You could look at the emit for the relation to determine what columns have to be duplicated and then add a dummy project which inserts the duplicated columns.

This is interesting - can you expand on this? (I think what you mean is to have output emitting like `[6, 7, 8, 9]` in this and handles that in the Substrait consumer but I am not sure)

> I think that would be fine for the by columns but misleading for the on columns. Due to the tolerance parameter the on columns are not necessarily strict duplicates.
True. The `on` is de duplicated because most of the time, people don't care about the time of the right side table (think it as a fuzzy alignment), and when they do - they will manually duplicate the right time column instead of relying the asof join node to do it. In other words, always de duplicate the time column is good enough for us and keeps the API simpler.

**Comment by westonpace** [comment]
> This is interesting - can you expand on this? (I think what you mean is to have output emitting like [6, 7, 8, 9] in this and handles that in the Substrait consumer but I am not sure)

Let's assume the inputs each have 1 `on` column, 1 `by` column, and 1 payload column.  The current asof join node will output 4 columns, `[on, by, p1, p2].  The Substrait plan will expect 6 columns, `[on1, by1, p1, on2, by2, p2]`.  As long as the emit from substrait is `[0, 1, 2, 5]` then we don't need to do anything.  If the emit is anything else then we can solve this by inserting an Acero project node to re-order and duplicate columns.

> True. The on is de duplicated because most of the time, people don't care about the time of the right side table (think it as a fuzzy alignment), and in the rare case that they do - they will manually duplicate the right time column instead of relying the asof join node to do it. In other words, always de duplicate the time column is good enough for us and keeps the API simpler.

If you don't mind duplicating the `on` column then this can be solved entirely in the substrait mapping layer by adding an extra project node to the acero plan, after the asof join node, if required, adding any duplicated columns that are not removed by substrait's emit.

**Comment by icexelloss** [comment]
> If you don't mind duplicating the on column then this can be solved entirely in the substrait mapping layer by adding an extra project node to the acero plan, after the asof join node, if required, adding any duplicated columns that are not removed by substrait's emit.

I am sure that I understand this 100% so I will try to rephrase:

Using the same example of `on1,by1,p1,on2,by2,p2` as the input columns, and `on1, by1, p1, p2` as the asof join output columns. Currently the emit mapping is `[6, 7, 8, 9]`. To make Acero work, we need to add a dummy projection after asof join in Acero that  `on1, by1, p1, p2` -> `on1, by1, p1, on1, by1, p2`, and then project it back to `on1, by1, p1, p2` as the user query specifies. By doing this, we can resolve the emit output mapping issue with the cost of dummy projection.

If that is correct, a follow up question is that does the dummy projection has any performance cost, i.e., does it need to copy the `on1` and `by1` column or does it can do it in a way that is 0-copy?

**Comment by icexelloss** [comment]
Also, reading this doc: https://substrait.io/relations/basics/#emit-output-ordering
It seems that there is `Remap` which states that "This remapping can be also used to drop columns no longer used (such as a filter field or join keys after a join). " - Does that not handle the case that we are talking about here?

**Comment by westonpace** [comment]
@icexelloss Here is an example of how this might be achieved.  It allows the extension full control to remap the emit info in whatever way makes sense.  For asof join the implementation was kind of complicated (for me) but it was just shuffling around numbers in the correct order :)

So, in the unit test (which matches the `[on1, by1, p1, on2, by2, p2]` case) the emit requested from the producer is `[0, 1, 2, 5]` and the extension creates the acero asof join node and rewrites the emit as `[0, 1, 2, 3]` which matches the Acero output.  If the user doesn't supply any emit then the extension creates the acero asof join node and rewrites the emit as `[0, 1, 2, 0, 1, 3]`.  The extension-independent substrait logic (e.g. `relation_internal.cc`) then takes this remapped emit info and inserts the appropriate dummy project node.

https://github.com/apache/arrow/commit/778c0c7cbedb0132a9a83f099f3bb97203372074

**Comment by icexelloss** [comment]
Update:

@westonpace @gforsyth and I had a meeting yesteday about this and we agreed on a fix for this issue.

The proposed fix is to include emit_output in the AsofJoin rel to handle the column de duplication, I have implemented it now the code produces a new plan:

```
extension_uris {
      extension_uri_anchor: 1
    }
    relations {
      root {
        input {
          project {
            common {
              emit {
                output_mapping: 6
                output_mapping: 7
                output_mapping: 8
                output_mapping: 9
              }
            }
            input {
              extension_multi {
                common {
                  emit {
                    output_mapping: 0
                    output_mapping: 1
                    output_mapping: 2
                    output_mapping: 5
                  }
                }
                inputs {
                  project {
                    common {
                      emit {
                        output_mapping: 3
                        output_mapping: 4
                        output_mapping: 5
                      }
                    }
                    input {
                      project {
                        common {
                          emit {
                            output_mapping: 3
                            output_mapping: 4
                            output_mapping: 5
                          }
                        }
                        input {
                          read {
                            base_schema {
                              names: "time"
                              names: "tid"
                              names: "price"
                              struct {
                                types {
                                  timestamp_tz {
                                    nullability: NULLABILITY_REQUIRED
                                  }
                                }
                                types {
                                  i32 {
                                    nullability: NULLABILITY_NULLABLE
                                  }
                                }
                                types {
                                  fp64 {
                                    nullability: NULLABILITY_NULLABLE
                                  }
                                }
                                nullability: NULLABILITY_REQUIRED
                              }
                            }
                            named_table {
                              names: "smooth:/research/group/tsmoto/bamboo/ci/test1?begin=1546300800000000000&end=1547078400000000000&columns=tid,price"
                            }
                          }
                        }
                        expressions {
                          selection {
                            direct_reference {
                              struct_field {
                              }
                            }
                            root_reference {
                            }
                          }
                        }
                        expressions {
                          selection {
                            direct_reference {
                              struct_field {
                                field: 1
                              }
                            }
                            root_reference {
                            }
                          }
                        }
                        expressions {
                          selection {
                            direct_reference {
                              struct_field {
                                field: 2
                              }
                            }
                            root_reference {
                            }
                          }
                        }
                      }
                    }
                    expressions {
                      selection {
                        direct_reference {
                          struct_field {
                          }
                        }
                        root_reference {
                        }
                      }
                    }
                    expressions {
                      selection {
                        direct_reference {
                          struct_field {
                            field: 1
                          }
                        }
                        root_reference {
                        }
                      }
                    }
                    expressions {
                      selection {
                        direct_reference {
                          struct_field {
                            field: 2
                          }
                        }
                        root_reference {
                        }
                      }
                    }
                  }
                }
                inputs {
                  project {
                    common {
                      emit {
                        output_mapping: 3
                        output_mapping: 4
                        output_mapping: 5
                      }
                    }
                    input {
                      project {
                        common {
                          emit {
                            output_mapping: 3
                            output_mapping: 4
                            output_mapping: 5
                          }
                        }
                        input {
                          read {
                            base_schema {
                              names: "time"
                              names: "tid"
                              names: "price"
                              struct {
                                types {
                                  timestamp_tz {
                                    nullability: NULLABILITY_REQUIRED
                                  }
                                }
                                types {
                                  i32 {
                                    nullability: NULLABILITY_NULLABLE
                                  }
                                }
                                types {
                                  fp64 {
                                    nullability: NULLABILITY_NULLABLE
                                  }
                                }
                                nullability: NULLABILITY_REQUIRED
                              }
                            }
                            named_table {
                              names: "smooth:/research/group/tsmoto/bamboo/ci/test1_daily?begin=1546300800000000000&end=1547078400000000000&columns=tid,price"
                            }
                          }
                        }
                        expressions {
                          selection {
                            direct_reference {
                              struct_field {
                              }
                            }
                            root_reference {
                            }
                          }
                        }
                        expressions {
                          selection {
                            direct_reference {
                              struct_field {
                                field: 1
                              }
                            }
                            root_reference {
                            }
                          }
                        }
                        expressions {
                          selection {
                            direct_reference {
                              struct_field {
                                field: 2
                              }
                            }
                            root_reference {
                            }
                          }
                        }
                      }
                    }
                    expressions {
                      selection {
                        direct_reference {
                          struct_field {
                          }
                        }
                        root_reference {
                        }
                      }
                    }
                    expressions {
                      selection {
                        direct_reference {
                          struct_field {
                            field: 1
                          }
                        }
                        root_reference {
                        }
                      }
                    }
                    expressions {
                      selection {
                        direct_reference {
                          struct_field {
                            field: 2
                          }
                        }
                        root_reference {
                        }
                      }
                    }
                  }
                }
                detail {
                  type_url: "[type.googleapis.com/arrow.substrait_ext.AsOfJoinRel](http://type.googleapis.com/arrow.substrait_ext.AsOfJoinRel)"
                  value: "\n\022\n\006\022\004\n\002\022\000\022\010\022\006\n\004\022\002\010\001\n\022\n\006\022\004\n\002\022\000\022\010\022\006\n\004\022\002\010\001\020\200\240\361\302\2614"
                }
              }
            }
            expressions {
              selection {
                direct_reference {
                  struct_field {
                  }
                }
                root_reference {
                }
              }
            }
            expressions {
              selection {
                direct_reference {
                  struct_field {
                    field: 1
                  }
                }
                root_reference {
                }
              }
            }
            expressions {
              selection {
                direct_reference {
                  struct_field {
                    field: 2
                  }
                }
                root_reference {
                }
              }
            }
            expressions {
              selection {
                direct_reference {
                  struct_field {
                    field: 3
                  }
                }
                root_reference {
                }
              }
            }
          }
        }
        names: "time"
        names: "tid"
        names: "price"
        names: "price_right"
      }
    }
``` 

Note that the emit info in the Asof Join node is now
```
                  emit {
                    output_mapping: 0
                    output_mapping: 1
                    output_mapping: 2
                    output_mapping: 5
                  }
```

However, the emit info the project following asof join seems off to me
```
              emit {
                output_mapping: 6
                output_mapping: 7
                output_mapping: 8
                output_mapping: 9
              }
```
If I understand correctly, with the emit info from the asof join node it's output schema now should have 4 columns instead of 6.

@westonpace has also asked a question about this via email which I will copy here:


```
As a result of this meeting we discovered that the asof join Ibis-Substrait extension needed changes.  Li Jin has made those changes and things still look wrong.

[@Gil Forsyth](mailto:gil@voltrondata.com) I am concerned about this logic in ibis-substrait: https://github.com/ibis-project/ibis-substrait/blob/main/ibis_substrait/compiler/translate.py#L778-L812

Ibis calculates `mapping_counter` as `itertools.count(sum(len(t.schema) for t in source_tables))`.  This `mapping_counter` is then used as the first emit index in the project relation.

Using source_tables here is very confusing and I'm not sure why that is happening.  This should be looking at the emitted output schema of the project input.  I don't see any reason to look at source_tables.  I would expect this to be incorrect even without asof join.  For example, if there were two projections back-to-back like I think this would generate:

r1 <- read(a, b, c): emit(default)
p1 <- project(r1, field(0), field(1)): emit(3, 4) # Correct
p2 <- project(p1, field(0)): emit(3) # Incorrect, should be emit(2)

I could be doing something wrong (and I don't know how to setup ibis-substrait to test) so I hope someone can check.
```

**Comment by gforsyth** [comment]
The `mapping_counter` code was a bit of spaghetti that I cleaned up in #465, but it does match the output outlined in @westonpace's example, e.g.

```
[ins] In [1]: import ibis
[ins] In [2]: t = ibis.table([("a", "string"), ("b", "float"), ("c", "int32")], "t")
[ins] In [3]: p1 = t.select("a", "b")
[ins] In [4]: p2 = p1.select("a")
[ins] In [5]: p2
Out[5]:
r0 := UnboundTable: t
  a string
  b float64
  c int32

r1 := Selection[r0]
  selections:
    a: r0.a
    b: r0.b

Selection[r1]
  selections:
    a: r1.a


[ins] In [6]: ibis.show_sql(p2)
SELECT
  t0.a
FROM (
  SELECT
    t1.a AS a,
    t1.b AS b
  FROM t AS t1
) AS t0

[ins] In [7]: from ibis_substrait.compiler.core import SubstraitCompiler
[ins] In [8]: compiler = SubstraitCompiler()
[ins] In [9]: out = compiler.compile(p2)

extension_uris {
  extension_uri_anchor: 1
}
relations {
  root {
    input {
      project {
        common {
          emit {
            output_mapping: 2
          }
        }
        input {
          project {
            common {
              emit {
                output_mapping: 3
                output_mapping: 4
              }
            }
            input {
              read {
                common {
                  direct {
                  }
                }
                base_schema {
                  names: "a"
                  names: "b"
                  names: "c"
                  struct {
                    types {
                      string {
                        nullability: NULLABILITY_NULLABLE
                      }
                    }
                    types {
                      fp64 {
                        nullability: NULLABILITY_NULLABLE
                      }
                    }
                    types {
                      i32 {
                        nullability: NULLABILITY_NULLABLE
                      }
                    }
                    nullability: NULLABILITY_REQUIRED
                  }
                }
                named_table {
                  names: "t"
                }
              }
            }
            expressions {
              selection {
                direct_reference {
                  struct_field {
                  }
                }
                root_reference {
                }
              }
            }
            expressions {
              selection {
                direct_reference {
                  struct_field {
                    field: 1
                  }
                }
                root_reference {
                }
              }
            }
          }
        }
        expressions {
          selection {
            direct_reference {
              struct_field {
              }
            }
            root_reference {
            }
          }
        }
      }
    }
    names: "a"
  }
}
```

Now, as for the `emit` following the `extension_multi` -- that is incorrect, as you suspect.
The `output_mapping` is calculated either as the sum of the parent tables (if there are no other `emits`) or we look at the most recent projection (https://github.com/ibis-project/ibis-substrait/blob/main/ibis_substrait/compiler/translate.py#L780-L786) by checking for `relation.project.common` but I think there needs to be an additional check for `relation.extension_multi.common` in order to pick up the `emit` from the `asof_join`.

Is `extension_multi` part of the `substrait` spec or is that a naming convention from the specific `asof_join` extension?

**Comment by gforsyth** [comment]
If it's part of the spec, we can add it to the `ops.Selection` translate code directly -- if not, I can refactor how we calculated `output_mapping` and make it into a `dispatcher` so that extensions could register a function for computing the `output_mapping` based on the type of `rel`.

**Comment by icexelloss** [comment]
`extension_multi` is part of the substrait spec: https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L281

Which the asof join is an implementation of

**Comment by gforsyth** [comment]
Great!  Let me put together a PR, then

**Comment by icexelloss** [comment]
I am curious why don't we always check the emit output mapping but have to do it on a node by node basis?

**Comment by westonpace** [comment]
> or we look at the most recent projection

> If it's part of the spec, we can add it to the ops.Selection translate code directly

This confuses me.  `emit` isn't really specific to projection.  What if you have:

read -> aggregate -> project

You would want `mapping_counter` to be based on the output schema of the aggregation right?

**Comment by gforsyth** [comment]
> You would want `mapping_counter` to be based on the output schema of the aggregation right?

Yep, just added that in #466
